Add CIS benchmark checks for generating a Kubernetes compliance report

stmcginnis opened the pull request:
**Issue number:**

Closes #2852 

**Description of changes:**

This adds a set of CIS benchmark checks to `bloodhound` for the Kubernetes CIS benchmark.

Still needs exposure through `apiserver` and `apiclient`. The plumbing for CIS reports in general are still being worked on for the CIS Bottlerocket benchmark. Depending on timing, the wiring up of this benchmark will either be added to this PR in additional commits, or added as a follow up PR after this one merges.

**Testing done:**

Ran new checks on `aws-k8s-1.26` node:

```
[ssm-user@control]$ apiclient report cis-k8s
Benchmark name:  CIS Kubernetes Benchmark (Worker Node)
Version:         v1.7.1
Reference:       https://www.cisecurity.org/benchmark/kubernetes
Benchmark level: 1
Start time:      2023-07-18T15:09:51.451720655Z

[PASS] 4.1.1     Ensure that the kubelet service file permissions are set to 644 or more restrictive (Automatic)
[PASS] 4.1.2     Ensure that the kubelet service file ownership is set to root:root (Automatic)
[SKIP] 4.1.3     If proxy kubeconfig file exists ensure permissions are set to 644 or more restrictive (Manual)
[SKIP] 4.1.4     If proxy kubeconfig file exists ensure ownership is set to root:root (Manual)
[PASS] 4.1.5     Ensure that the --kubeconfig kubelet.conf file permissions are set to 644 or more restrictive (Automatic)
[PASS] 4.1.6     Ensure that the --kubeconfig kubelet.conf file ownership is set to root:root (Automatic)
[PASS] 4.1.7     Ensure that the certificate authorities file permissions are set to 600 or more restrictive (Automatic)
[PASS] 4.1.8     Ensure that the client certificate authorities file ownership is set to root:root (Automatic)
[PASS] 4.1.9     If the kubelet config.yaml configuration file is being used validate permissions set to 600 or more restrictive (Automatic)
[PASS] 4.1.10    If the kubelet config.yaml configuration file is being used validate file ownership is set to root:root (Automatic)
[PASS] 4.2.1     Ensure that the --anonymous-auth argument is set to false (Automatic)
[PASS] 4.2.2     Ensure that the --authorization-mode argument is not set to AlwaysAllow (Automatic)
[PASS] 4.2.3     Ensure that the --client-ca-file argument is set as appropriate (Automatic)
[PASS] 4.2.4     Verify that the --read-only-port argument is set to 0 (Automatic)
[PASS] 4.2.5     Ensure that the --streaming-connection-idle-timeout argument is not set to 0 (Automatic)
[PASS] 4.2.6     Ensure that the --make-iptables-util-chains argument is set to true (Automatic)
[SKIP] 4.2.7     Ensure that the --hostname-override argument is not set (not valid for Bottlerocket) (Manual)
[PASS] 4.2.9     Ensure that the --tls-cert-file and --tls-private-key-file arguments are set as appropriate (Automatic)
[SKIP] 4.2.10    Ensure that the --rotate-certificates argument is not set to false (not valid for Bottlerocket) (Manual)
[PASS] 4.2.11    Verify that the RotateKubeletServerCertificate argument is set to true (Automatic)
[PASS] 4.2.12    Ensure that the Kubelet only makes use of Strong Cryptographic Ciphers (Automatic)
[PASS] 4.2.13    Ensure that a limit is set on pod PIDs (Automatic)

Passed:          18
Failed:          0
Skipped:         4
Total checks:    22

Compliance check result: PASS
```
```
[ssm-user@control]$ apiclient report cis-k8s -l 2
Benchmark name:  CIS Kubernetes Benchmark (Worker Node)
Version:         v1.7.1
Reference:       https://www.cisecurity.org/benchmark/kubernetes
Benchmark level: 2
Start time:      2023-07-18T15:10:15.191556058Z

[PASS] 4.1.1     Ensure that the kubelet service file permissions are set to 644 or more restrictive (Automatic)
...
```

```
[ssm-user@control]$ apiclient report cis-k8sh
Missing or unknown subcommand for 'report'
...
```

```json
apiclient report cis-k8s -l 2 -f json
{
  "level": 2,
  "total": 23,
  "passed": 18,
  "skipped": 5,
  "failed": 0,
  "status": "PASS",
  "timestamp": "2023-07-18T15:10:32.352887194Z",
  "name": "CIS Kubernetes Benchmark (Worker Node)",
  "version": "v1.7.1",
  "url": "https://www.cisecurity.org/benchmark/kubernetes",
  "results": {
    "k8s04010100": {
      "name": "k8s04010100",
      "id": "4.1.1",
      "level": 1,
      "title": "Ensure that the kubelet service file permissions are set to 644 or more restrictive",
      "mode": "Automatic",
      "status": "PASS",
      "error": ""
    },
   ...
}
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

stmcginnis committed:
bloodhound: Add base Kubernetes checker binary

This adds a multicall binary to be used as the common entry point for
all Kubernetes CIS benchmark checks.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.1.1 check

This adds the 4.1.1 check to verify the kubelet service file has
restrictive permissions.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.1.2 check

This adds the 4.1.2 check to make sure the kubelet service file
ownership is root:root.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.1.5 check

This adds the 4.1.5 check to ensure the kubelet.conf file has
restrictive permissions.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.1.6 check

This adds the 4.1.6 check to verify kubelet.conf is owned by root:root.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.1.7 check

This adds the 4.1.7 check to verify the kubelet CA file has restrictive
permissions.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.1.8 check

This adds check 4.1.8 to verify the kubelet CA file is owned by
root:root.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.1.9 check

This adds the 4.1.9 check to verify the kubelet configuration file has
restrictive permissions.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.1.10 check

This adds check 4.1.10 to verify the kubelet config file is owned by
root:root.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.2.1 check

This adds check 4.2.1 to verify anonymous auth is not enabled.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.2.2 check

This adds check 4.2.2 to verify authorization mode is not set to
AlwaysAllow.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.2.3 check

This adds check 4.2.3 to verify clientCAFile is configured and points to
an actual file.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.2.4 check

This adds the 4.2.4 to verify the readOnlyPort is not set to 0.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.2.5 check

This adds check 4.2.5 to verify streamingConnectionIdleTimeout is not
set to 0.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.2.6 check

This adds check 4.2.6 to verify makeIPTablesUtilChains is true.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.2.9 check

This adds the check to make sure the server TLS certificates are
configured correctly.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.2.10 check

This adds check 4.2.10 to verify rotateCertificates is not disabled.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.2.11 check

This adds check 4.2.11 to verify RotateKubeletServerCertificate is not
disabled.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Add Kubernetes 4.2.12 check

This adds check 4.2.12 to check that kubelet is only configured to use
secure cryptographic ciphers.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

maxres-ch commented:
This is great to see! I wrote a daemonset to run the bash script outlined here: https://github.com/aws-samples/containers-blog-maelstrom/tree/main/cis-bottlerocket-benchmark-eks/

Will there be an operator to accompany this or will it be up to users to invoke this client and report the status individually?

stmcginnis commented:
> Will there be an operator to accompany this or will it be up to users to invoke this client and report the status individually?

Hey! Great questions. The plan is for the Bottlerocket and Kubernetes benchmarks to ultimately be exposed via `apiclient`. That work is still in progress, with the `apiserver` updates happening [here](https://github.com/bottlerocket-os/bottlerocket/pull/3219), and the last step of wiring in to `apiclient` coming once that is done.

So while it's possible to execute this command directly to get the output, the actual end user experience should be a little cleaner through `apiclient report cis-k8s` (or something like that).

bcressey commented on sources/bloodhound/src/bin/kubernetes-checks/main.rs:
```diff
@@ -0,0 +1,54 @@
+            title: "Ensure that the --hostname-override argument is not set".to_string(),
+            id: "4.2.8".to_string(),
+            level: 1,
+        }),
```
For these manual checks, is the logic that we know that they'll potentially fail on Bottlerocket, so we turn them into "SKIP" tests? E.g. we always set `hostname-override` now to placate the external AWS cloud provider.

stmcginnis replied:
Hmm... for the most part, these manual checks are supposed to be things where the benchmark defines them as manual and/or we do not have a clear way of verifying things in an automated way. So it's an indication that the user will have to perform the validation themselves to make sure it is correct for their given environment and needs.

For this particular one, I can't recall now why it is manual. We have a mismatch here where the K8s CIS benchmark says this should not be overridden, but in Bottlerocket's case we have unique constraints where it absolutely _should_ be overridden.

I see a few ways we could handle this:

1. Add something like "(not applicable for Bottlerocket)" to the title to show it can be ignored.
2. Just have the checker return `PASS`. Seems like lying, but we know in our case it is OK to be doing it this way.
3. Exclude completely from the set of checks. Could lead to questions of why the built in checkers are incomplete and cause some doubt about trustworthiness of our results, but would side step the fact that it does not apply to Bottlerocket.

Any preferences or other ideas?

bcressey replied:
The "(not valid for Bottlerocket)" notes look good to me.

bcressey commented on sources/bloodhound/src/bin/kubernetes-checks/checks.rs:
```diff
@@ -0,0 +1,27 @@
+
+impl Checker for K8S04010100Checker {
+    fn execute(&self) -> CheckerResult {
+        check_file_not_mode(KUBELET_SERVICE_FILE, 0b111111)
```
It looks like this ends up only checking `0o77` rather than `0o177`, so we wouldn't detect a `744` file.

It might be better to compose it out of libc constants, e.g.
```
use libc::{S_IRUSR, S_IWUSR};

let u_rw_g_o_ = S_IRUSR | S_IWUSR;
let mask = !u_rw_g_o_ & 0o777;
```

stmcginnis replied:
Updated - that does make it a little more explicitly clear what we are looking for. 👍

bcressey commented on sources/bloodhound/src/lib.rs:
```diff
@@ -151,6 +153,24 @@ pub fn check_file_not_mode(file_path: &str, mode: u32) -> CheckerResult {
     result
 }
 
+/// Verifies the file at the given path is owned by root:root, returning a `Results` based on the results.
+pub fn ensure_file_owner_and_group_root(file_path: &str) -> CheckerResult {
```
nit: why not take a desired uid and gid? That seems more consistent with the mode checker which takes a mask rather than hard-coding the specific check.

stmcginnis replied:
Mainly because all we ever care about with the benchmarks is root ownership, and there wouldn't ever be a need to check any other arbitrary uid/gid. So this keeps it simple and doesn't add extra parameters that would never be used.

If you feel strongly about enabling that ability, let me know and I can change it. Or we can also wait if/until it ever comes up that we have multiple uid/gid's to deal with on Bottlerocket.

bcressey commented on sources/bloodhound/src/bin/kubernetes-checks/checks.rs:
```diff
@@ -0,0 +1,714 @@
+
+    fn metadata(&self) -> CheckerMetadata {
+        CheckerMetadata {
+            title: "Ensure that the kubelet service file permissions are set to 644 or more restrictive".to_string(),
```
The v1.7.0 benchmark seems to use this for all permissions:
```suggestion
            title: "Ensure that the kubelet service file permissions are set to 600 or more restrictive".to_string(),
```

stmcginnis replied:
This was changed in the 1.7.1 benchmark.

bcressey commented on sources/Cargo.lock:
```diff
@@ -4297,6 +4333,12 @@ version = "0.1.11"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "e1766d682d402817b5ac4490b3c3002d91dfa0d22812f341609f97b08757359c"
 
+[[package]]
+name = "unsafe-libyaml"
+version = "0.2.8"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "1865806a559042e51ab5414598446a5871b561d21b6764f2eabb0dd481d880a6"
```
I'm not very excited about adding libyaml to the distro given its security track record. Do we have other options?

stmcginnis replied:
Not that I have found, but I will continue to look in to any alternatives. Really don't want to manually parse the yaml config file, but that's always an option too if this becomes problematic.

stmcginnis replied:
The only other reasonable yaml library I could find was `yaml_rust`, but that appears to be no longer maintained.

bcressey replied:
On closer inspection, `unsafe-libyaml` is a fork of libyaml translated into unsafe Rust.

We can keep an eye on it, but it seems OK for now.

bcressey commented on sources/bloodhound/src/bin/kubernetes-checks/checks.rs:
```diff
@@ -228,3 +228,51 @@ impl Checker for K8S04020100Checker {
+
+impl Checker for K8S04020200Checker {
+    fn execute(&self) -> CheckerResult {
+        #[derive(Deserialize)]
+        struct Authorization {
+            mode: String,
+        }
+
+        #[derive(Deserialize)]
+        struct KubeletConfig {
+            authorization: Authorization,
+        }
```
I don't mind this very much, but there'd be less duplicated code to have a separate `KubeletConfig` struct with all the values for all the checks. Maybe a final refactor commit on top of these individual ones to pull it out.

stmcginnis replied:
Yeah... the tricky part here is that we actual do expect some of these to fail to deserialize in most cases. So that makes it a challenge to have one unifying struct defined. A missing property only relevant to one check would cause a deserialization error for another check.

We could do some custom deserialization and use things like `Option<x>`, but then that does add extra complexity where the result would probably end up being more code rather than less. It's a little messier, but I do like how now it is very localized and clear what is being looked at in each case.

But if you feel strongly about it, let me know and I can spend some time looking at how to handle a single definition to use for all `KubeletConfig` settings.

bcressey commented on sources/bloodhound/src/bin/kubernetes-checks/checks.rs:
```diff
@@ -419,3 +419,46 @@ impl Checker for K8S04020500Checker {
+            title: "Ensure that the --protect-kernel-defaults argument is set to true".to_string(),
+            id: "4.2.6".to_string(),
+            level: 1,
+            name: "k8s04020600".to_string(),
```
In the 1.7.0 benchmark, the `--protect-kernel-defaults` recommendation no longer exists, and the numbering of subsequent recommendations shifts upward.

stmcginnis replied:
Well that's... unfortunate.

bcressey commented on sources/bloodhound/src/bin/kubernetes-checks/checks.rs:
```diff
@@ -555,3 +555,47 @@ impl Checker for K8S04021000Checker {
+                }
+            } else {
+                // Normally this value should not be present in the config file, so deserialization is expected to fail.
+                result.status = CheckStatus::PASS;
```
`rotateCertificates` defaults to false (disabled) - so if it's not present in the config then it's a failed check.

This seems like the hostname override recommendation where it doesn't really apply to EKS worker nodes - but might apply for other cloud providers.

stmcginnis replied:
Double checked documentation. There is a page that states `kubeadm` sets this to true by default. That mis-led me. Verified in the `kubelet` config docs this is indeed `false` by default. Good catch!

bcressey commented on sources/bloodhound/src/bin/kubernetes-checks/checks.rs:
```diff
@@ -649,3 +649,66 @@ impl Checker for K8S04021200Checker {
+        if let Ok(kubelet_file) = File::open(KUBELET_CONF_FILE) {
+            if let Ok(config) = serde_yaml::from_reader::<_, KubeletConfig>(kubelet_file) {
+                let mut allowed = true;
+                for suite in config.tls_cipher_suites {
+                    if !allowed_suites.iter().any(|&i| i == suite) {
+                        allowed = false;
+                        result.error = format!("'{}' not an allowed cipher suite", suite);
+                        break;
+                    }
+                }
```
the lists are small, but this would avoid the O(nm) algorithm:
```
let configured_suites = HashSet::from(config.tls_cipher_suites);
if !configured_suites.is_subset(allowed_suites) {
    result.error = format!("Found disallowed cipher suites");
}
```

stmcginnis replied:
Done!

bcressey commented on sources/bloodhound/src/bin/kubernetes-checks/checks.rs:
```diff
@@ -0,0 +1,714 @@
+        result
+    }
+
+    fn metadata(&self) -> CheckerMetadata {
+        CheckerMetadata {
+            title: "Ensure that the Kubelet only makes use of Strong Cryptographic Ciphers"
+                .to_string(),
+            id: "4.2.13".to_string(),
+            level: 1,
+            name: "k8s04021300".to_string(),
+            mode: Mode::Automatic,
+        }
+    }
```
We are missing the current `4.2.13` check - "Ensure that a limit is set on pod PIDs (Manual)".

stmcginnis replied:
Added! It's in the benchmark as a manual check, but easy enough for us to validation automatically for Bottlerocket. 

Note: this check will fail until https://github.com/bottlerocket-os/bottlerocket/pull/3263 is merged.

stmcginnis committed:
bloodhound: Add Kubernetes 4.2.13 check

This adds verification that podPidsLimit is configured to restrict the
maximum number of PIDs allowed per-pod.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
bloodhound: Skip K8s 4.2.10 as not applicable

This changes the 4.2.10 check to be a manual test response. This check
does not apply to EKS or EKS-D since external IAM auth is used.

This leaves the validation logic in place as we may want to use it in
the future. For not EKS-related use cases, it may make sense to enable
configuration of certificate rotation. If/when that is done, this check
can be reenabled and updated to validate under the proper conditions.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
apiserver: Add support for Kuberenetes CIS report

This adds handling for a "type" parameter to be passed along to the
`/report/cis` endpoint to control what type of CIS report to generate.
The default type is the Bottlerocket CIS benchmark report, with
"kubernetes" being the only recognized other option.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
apiclient: Add `report cis-k8s` command

This adds the `cis-k8s` report subcommand to access the Kubernetes CIS
report. It uses the standard CIS arguments of `-l` for level and `-f`
for format.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis commented:
Added `apiserver` and `apiclient` parts of the complete implementation and rebased on `develop` to resolve conflicts.

arnaldo2792 commented on sources/api/apiclient/README.md:
````diff
@@ -193,13 +193,9 @@ For example, if you want the name "FOO", you can `PATCH` to `/settings?tx=FOO` a
-```shell
-apiclient report cis
-```
+These set of reports can be used to evaluate the current system state and settings for compliance with the the [Center for Internet Security (CIS) Benchmarks].
````
```suggestion
These set of reports can be used to evaluate the current system state and settings for compliance with the [Center for Internet Security (CIS) Benchmarks].
```

arnaldo2792 commented on sources/api/apiclient/README.md:
```diff
@@ -211,31 +207,68 @@ By default, Bottlerocket instances meet level 1 compliance unless specific setti
 
+**Format**
+
+The default format of the CIS reports are a human readable text report.
```
```suggestion
The default format of the CIS reports is a human readable text report.
```

stmcginnis replied:
Oh right, singular "default format". ;)

arnaldo2792 commented on sources/api/apiclient/README.tpl:
````diff
@@ -193,13 +193,9 @@ For example, if you want the name "FOO", you can `PATCH` to `/settings?tx=FOO` a
-```shell
-apiclient report cis
-```
+These set of reports can be used to evaluate the current system state and settings for compliance with the the [Center for Internet Security (CIS) Benchmarks].
````
```suggestion
These set of reports can be used to evaluate the current system state and settings for compliance with the [Center for Internet Security (CIS) Benchmarks].
```

arnaldo2792 commented on sources/api/apiclient/src/main.rs:
```diff
@@ -151,6 +152,7 @@ fn usage() -> ! {
             reboot                     Reboots the host.
             exec                       Execute a command in a host container.
             report cis                 Retrieve a Bottlerocket CIS benchmark compliance report.
+            report cis-k8s             Retrieve a Kubernetes CIS benchmark compliance report.
```
Shouldn't this command be “conditionality”-compiled based on the variant?

stmcginnis replied:
I did think about that and decided to just include it. If someone were to try to run the report on an `aws-ecs-*` instance, it would just report back SKIPPED for every assertion.

Conditionally including it makes things slightly more complicated with needing conditional inclusion, especially here with things like documentation. It also changes this CLI "api" by having some commands there on one instance, but not there on others.

It could be done, but I opted to try to keep things as simple as possible. Willing to hear strong opinions though.

arnaldo2792 commented on sources/bloodhound/src/bin/kubernetes-checks/checks.rs:
```diff
@@ -0,0 +1,724 @@
+
+    fn metadata(&self) -> CheckerMetadata {
+        CheckerMetadata {
+            title: "If the kubelet config.yaml configuration file is being used validate permissions set to 600 or more restrictive".to_string(),
```
It seems like all the `kubelet` systemd services do pass `--config <path>`, so could this be?:

```suggestion
            title: "Ensure --config config.yaml file permissions are set to 600 or more restrictive".to_string(),
```

stmcginnis replied:
See above, this is the title of the check.

arnaldo2792 commented on sources/bloodhound/src/bin/kubernetes-checks/checks.rs:
```diff
@@ -0,0 +1,724 @@
+
+    fn metadata(&self) -> CheckerMetadata {
+        CheckerMetadata {
+            title: "If the kubelet config.yaml configuration file is being used validate file ownership is set to root:root"
```
Same as in check `K8S04010900Checker`

arnaldo2792 commented on sources/bloodhound/src/bin/kubernetes-checks/checks.rs:
```diff
@@ -0,0 +1,724 @@
+
+    fn metadata(&self) -> CheckerMetadata {
+        CheckerMetadata {
+            title: "Ensure that the kubelet service file permissions are set to 644 or more restrictive".to_string(),
```
Is the `title` intended to be use as an instruction for the user? Or as a description for the checker? If the latter, then shouldn't `Ensure` be `Ensures`? Otherwise, it sounds like we are instructing the users they should set the correct permissions.

stmcginnis replied:
Neither? The title is just the title of the check from the benchmark.

arnaldo2792 commented on sources/bloodhound/src/bin/kubernetes-checks/checks.rs:
```diff
@@ -0,0 +1,724 @@
+                {
+                    result.status = CheckStatus::PASS;
+                } else {
+                    result.error = "TLS files not set to expected path".to_string();
```
Should the expected path be included in this error?

stmcginnis replied:
It is dependent on what is in the config file, which may not contain anything. So this is a general message to point at where the issue is, and the benchmark contains the steps necessary to manually verify if these values are not found in the config file where they are expected to be.

arnaldo2792 commented on sources/bloodhound/src/bin/kubernetes-checks/checks.rs:
```diff
@@ -0,0 +1,724 @@
+                    .map(|s| s.as_str())
+                    .collect();
+                if !configured_suites.is_subset(&allowed_suites) {
+                    result.error = "Found disallowed cipher suites".to_string();
```
Should the culprits be included in this error?

stmcginnis replied:
This is another case where we would want them to refer to the benchmark to evaluate things. This would need to be something they are doing in a custom build, since we know the official Bottlerocket variants do not include any cipher suites that are not acceptable.

bcressey commented on sources/api/apiclient/README.tpl:
````diff
@@ -193,13 +193,9 @@ For example, if you want the name "FOO", you can `PATCH` to `/settings?tx=FOO` a
-```shell
-apiclient report cis
-```
+These set of reports can be used to evaluate the current system state and settings for compliance with the [Center for Internet Security (CIS) Benchmarks].
````
```suggestion
These reports can be used to evaluate the current system state and settings for compliance with the [Center for Internet Security (CIS) Benchmarks].
```

(Or "This set of reports")

bcressey commented on sources/bloodhound/src/lib.rs:
```diff
@@ -288,4 +340,70 @@ mod test_utils {
+            .display()
+            .to_string();
+
+        // Then us that name to create test file with the expected permission mode
```
nit:
```suggestion
        // Then use that name to create test file with the expected permission mode
```

bcressey commented on sources/bloodhound/src/lib.rs:
```diff
@@ -288,4 +340,70 @@ mod test_utils {
+
+    #[test]
+    fn test_check_file_not_mode_file_not_found() {
+        // Generate a temp file that will be dropped as unused
```
nit: comment doesn't apply
```suggestion
```

grodriguezl commented:
This is beautiful. Thanks!
